pkg/controllers/vrep/replication_reconciler.go

```diff
@@ -0,0 +1,311 @@
+		return res, err
+	}
+	r.SourceVdb = sourceVdb
+	// fetch target vdb
+	targetVdb := &vapi.VerticaDB{}
+	targetName := names.GenNamespacedName(r.Vrep, r.Vrep.Spec.Target.VerticaDB)
+	if res, err = vk8s.FetchVDB(ctx, r.VRec, r.Vrep, targetName, targetVdb); verrors.IsReconcileAborted(res, err) {
+		return res, err
+	}
+	r.TargetVdb = targetVdb
+
+	return
+}
+
+// makeDispatcher will create a Dispatcher object based on the feature flags set.
+func (r *ReplicationReconciler) makeDispatcher() error {
+	if vmeta.UseVClusterOps(r.SourceVdb.Annotations) {
+		r.dispatcher = vadmin.MakeVClusterOps(r.Log, r.SourceVdb,
+			r.VRec.GetClient(), r.SourcePassword, r.VRec, vadmin.SetupVClusterOps)
+		return nil
+	}
+	return fmt.Errorf("replication is not supported for source VerticaDB with admintools deployment")
+}
+
+func (r *ReplicationReconciler) determineDBNames() {
+	r.TargetDBName = r.TargetVdb.Spec.DBName
+}
+
+func (r *ReplicationReconciler) determineUsernameAndPassword(ctx context.Context) (err error) {
+	r.SourceUsername, r.SourcePassword, err = setUsernameAndPassword(ctx,
+		r.Client, r.Log, r.VRec, r.SourceVdb, &r.Vrep.Spec.Source)
+	if err != nil {
+		return err
+	}
+
+	r.TargetUsername, r.TargetPassword, err = setUsernameAndPassword(ctx,
+		r.Client, r.Log, r.VRec, r.TargetVdb, &r.Vrep.Spec.Target)
+	if err != nil {
+		return err
+	}
+
+	return
+}
+
+func setUsernameAndPassword(ctx context.Context, cli client.Client, log logr.Logger,
+	vRec *VerticaReplicatorReconciler, vdb *vapi.VerticaDB,
+	dbInfo *v1beta1.VerticaReplicatorDatabaseInfo) (username, password string, err error) {
+	if dbInfo.UserName == "" {
+		// database superuser is assumed
+		username := vdb.GetVerticaUser()
+		password, err := vk8s.GetSuperuserPassword(ctx, cli, log, vRec, vdb)
+		if err != nil {
+			return "", "", err
+		}
+		return username, password, nil
+	} else {
+		// custom username and password is used
+		username := dbInfo.UserName
+		if dbInfo.PasswordSecret == "" {
+			// empty password is assumed
+			return username, "", nil
+		} else {
+			// fetch custom password
+			// assuming the password secret key is default
+			password, err := vk8s.GetCustomSuperuserPassword(ctx, cli, log,
+				vRec, vdb, dbInfo.PasswordSecret, "")
+			if err != nil {
+				return "", "", err
+			}
+			return username, password, nil
+		}
+	}
+}
+
+func (r *ReplicationReconciler) collectPodFacts(ctx context.Context) (err error) {
+	r.SourcePFacts, err = r.makePodFacts(ctx, r.SourceVdb,
+		r.Vrep.Spec.Source.SandboxName)
+	if err != nil {
+		return
+	}
+	if err = r.SourcePFacts.Collect(ctx, r.SourceVdb); err != nil {
+		return
+	}
+
+	r.TargetPFacts, err = r.makePodFacts(ctx, r.TargetVdb,
+		r.Vrep.Spec.Target.SandboxName)
+	if err != nil {
+		return
+	}
+	if err = r.TargetPFacts.Collect(ctx, r.TargetVdb); err != nil {
+		return
+	}
+	return
+}
+
+func (r *ReplicationReconciler) determineSourceAndTargetHosts() (err error) {
+	// assume source must not be read-only, no subcluster constraints
```

Comment:
I can relax it to allow read-only for source. Still not 100% sure it will work for read-only source. We may be able to add some tests (e2e?) later for this.

Reply:
It's a hard one to test because to get read-only nodes you have to kill pods. And then the operator will want to step in and restart pods anyway. Maybe that's an argument to assume source cannot be read-only. We should include that as a comment in case anyone wants to try and optimize this in the future.